### PR TITLE
[Core] Allow to pass node ip address to gcs server.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Some users (actually, most of the users who use private clusters I've seen) reported that their CPP address resolution didn't work, but the python one does. I assume our Python code is much more robust to resolve addresses. This PR allows Ray to pass the node IP address to the GCS server. 

## Related issue number

https://github.com/ray-project/ray/issues/10574

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
@raoul-khour-ts Can you try this PR and see if this works?

File: src/ray/gcs/gcs_server/gcs_server.h [hunk 1]
```diff
@@ -34,6 +34,7 @@ struct GcsServerConfig {
   uint16_t redis_port = 6379;
   bool retry_redis = true;
   bool is_test = false;
+  std::string node_ip_address = "127.0.0.1";
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_server.h:37]: If the user sets node_ip_address to 127.0.0.1, it does not work? Maybe the default value can be empty.

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_server.h:37]: Good point. Will fix.

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 2]
```diff
@@ -252,11 +252,13 @@ std::unique_ptr<GcsObjectManager> GcsServer::InitObjectManager() {
 }
 
 void GcsServer::StoreGcsServerAddressInRedis() {
-  std::string address =
-      GetValidLocalIp(
-          GetPort(),
-          RayConfig::instance().internal_gcs_service_connect_wait_milliseconds()) +
-      ":" + std::to_string(GetPort());
+  std::string ip = config_.node_ip_address;
+  if (ip == "") {
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_server.cc:256]:
```suggestion
  if (ip.empty()) {
```

File: src/ray/gcs/gcs_server/gcs_server.h [hunk 3]
```diff
@@ -34,6 +34,7 @@ struct GcsServerConfig {
   uint16_t redis_port = 6379;
   bool retry_redis = true;
   bool is_test = false;
+  std::string node_ip_address = "";
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_server.h:37]:
```suggestion
  std::string node_ip_address;
```

**Review by ffbin — APPROVED** [review]
LGTM

**Comment by roireshef** [comment]
HI @rkooo567 - What versions and setups are impacted by this issue?
More specifically, if I use python version of Ray v0.8.7 - is it supposed to work?
I'm double-checking since in https://github.com/ray-project/ray/issues/10574 it says that 0.8.7 should work and only 0.9.0 isn't. Did I miss anything?

**Comment by raoul-khour-ts** [comment]
@roireshef Sorry I was comparing v0.8.7 to v1.0.0 (I didn't realize 0.9.0dev at the time was actually 1.0.0). But yeah, this issue still happens on 1.0.0 (but is gone in 1.1.0dev). Should only effect people that are working in a private network that does not have an open proxy to 8.8.8.8

**Comment by roireshef** [comment]
Hi @raoul-khour-ts I'm running an experiment (migration to 0.8.7 from older version) on a private network, on top of docker containers. I've ran:
1. On the head node:
```
ray start --block --head --port=$redis_port --redis-password=$redis_password --node-ip-address=$head_node_ip \
--gcs-server-port=6005 --dashboard-port=6006 --node-manager-port=6007 --object-manager-port=6008 \
--redis-shard-ports=6400,6401,6402,6403,6404,6405,6406,6407,6408,6409 --min-worker-port=6100 --max-worker-port=6299
```
2. On the worker node:
```
ray start --block --address=$redis_address --redis-password=$redis_password --node-ip-address=$worker_node_ip 
--node-manager-port=6007 --object-manager-port=6008 --min-worker-port=6100 --max-worker-port=6299
```
3. So far everything looks to be working. On the head node I'm opening a new python shell and running:
```
import ray
ray.init(address='auto', redis_password='5241590000000000')
ray.cluster_resources()
```

And what I see are only the resources of the head node. That is in contrast to what the dashboard shows when I'm looking at it with my browser - it actually shows both nodes with their resources listed.

So my question is, since I'm using python and you wrote above that it shouldn't be affected, is there a hotfix I can apply to the Ray code to make 0.8.7 work? 

I'm currently blocked, and would sincerely appreciate your help!

**Comment by raoul-khour-ts** [comment]
Hey @roireshef 
So python is effected.

A solution that I have used to get 0.8.7 to work is:



ray start --head --gcs-server-port=ddd --cpus=0
```python
import redis
r = redis.Redis(host=xxx, port=yyy, password=zzz, db=0)
r.get('GcsServerAddress')
>>> 127.0.0.1:ddd
r.set('GcsServerAddress', xxx:ddd)
```

**Pull request merged**